### DISCUSSION TITLE
Consider using PicoCLI ?

### DISCUSSION THREAD

**Discussion by aadrian** [babashka/babashka#773] [Ideas]
Hi,

Any chance considering Pico CLI https://picocli.info/  for BB ?

IMO it's the most polished and well maintained CLI solution for the JVM ( but compared to non-JVM too).

Just a few points out of the box:
 - very detailed documentation
 - ansi color output
 - tab completion
 - sub-commands
 - GraalVM aware

Has a very small footprint so it's being used in many native CLI tools (thanks to GraalVM https://picocli.info/picocli-on-graalvm.html ) .

**Comment by borkdude** [comment]
@aadrian Can you give a few examples of how to use PicoCLI in Clojure? AFAIK it's not very well known in the Clojure community yet?

**Reply by aadrian** [reply]
> Can you give a few examples of how to use PicoCLI in Clojure?

@borkdude I'm sorry :( - but I don't have good examples of using PicoCLI **in** Clojure - see old issues https://github.com/remkop/picocli/issues/231  .
For the moment I'm using mostly PicoCLI **with** Clojure _(and other JVM based languages)_.

Mostly for simplicity and practicality, my JVM projects are mixed, i.e. also have Java files - mostly Main classes, sometimes utility classes (Gradle makes it easy to mix - e.g. for Clojure  https://clojurephant.dev/clojurephant/ ) .

**Reply by borkdude** [reply]
@aadrian Then how would it be useful for users to add this library to babashka?

**Reply by aadrian** [reply]
> Then how would it be useful for users to add this library to babashka?

@borkdude 

Some of the useful things PicoCLI could add (or handle for BB):
 - ansi color output for `bb --help`
 - nice and effortless help output as content by aggregating automaticallyexisting functionality (not separately written and out of sync)
 - tab completion in bash for the all the `bb` options and parameters
 - validation of options and parameters
 
just to mention a few (and the GraalVM footprint of that it is very small).

*I'm getting all the above in all CLI projects with little effort, just by using that library - IMHO it could enhance `bb`  too.*

**Reply by borkdude** [reply]
@aadrian OK, so only for bb itself and not for writers of bb scripts. Could you show me a couple of examples of your projects that I can readily use on the command line?
